### PR TITLE
show filename when open() fails

### PR DESCRIPTION
**Pull request by gfx**
`File.open()` does not show filename on errors, i.e.:

```
mirb> File.open("foo")
(mirb):1:open (RuntimeError)
```

I think filename is inevitable to debug mruby scripts, so I add the information to messages in this PR:

```
mirb> File.open("foo")
(mirb):1:open foo (RuntimeError)
```

I think the patch is a bit messy; I'll write some patch to mruby for `mb_sys_fail(mrb, format, ...)`, but  even though such a mruby patch, this PR is valuable for portability.

### CODE REVIEW

File: src/io.c [hunk 1]
```diff
@@ -39,6 +40,7 @@
 #include <stdio.h>
 #include <string.h>
 
+mrb_value mrb_vformat(mrb_state *mrb, const char *format, va_list ap); // not exported yet
```
Comment (gfx) [inline src/io.c]: It should be exported but not :(

File: src/io.c [hunk 2]
```diff
@@ -92,6 +94,35 @@ io_set_process_status(mrb_state *mrb, pid_t pid, int status)
 }
 #endif
 
+// vargs version of mrb_sys_fail()
+static
+void mrb_io_fail(mrb_state *mrb, const char *format, ...)
```
Comment (gfx) [inline src/io.c]: This is almost a copy of mrb_sys_fail()

cf. https://github.com/mruby/mruby/blob/master/src/error.c#L490

**Comment by tsahara** [comment]
Thank you for reporting it.  The patch looks good but it can be much simpler.

```
diff --git a/src/io.c b/src/io.c
index dd030b8..a4a209c 100644
--- a/src/io.c
+++ b/src/io.c
@@ -576,7 +576,7 @@ reopen:
         goto reopen;
       }
     }
-    mrb_sys_fail(mrb, "open");
+    mrb_sys_fail(mrb, pathname);
   }
 
   if (fd <= 2) {
```

With this change, mruby shows:

```
% mruby -e 'File.open("foo")'
trace:
        [0] /d/home/tsahara/src/mruby/mruby-io/mrblib/file.rb:14:in File.initialize
        [1] /d/home/tsahara/src/mruby/mruby-io/mrblib/io.rb:15:in IO#open
        [2] -e:1
/d/home/tsahara/src/mruby/mruby-io/mrblib/file.rb:14:No such file or directory - foo (Errno::ENOENT)
```

And this is how CRuby (2.3.1) shows:

```
% ruby -e 'File.open("foo")'
-e:1:in `initialize': No such file or directory @ rb_sysopen - foo (Errno::ENOENT)
        from -e:1:in `open'
        from -e:1:in `<main>'
```

**Comment by gfx** [comment]
I just clone this repo and the backtrace doesn't include the method name except for the message itself:

```
$ tmp/mruby/bin/mruby -e 'File.open("foo")'
trace:
	[0] -e:1
-e:1:open (RuntimeError)
```


How can I see the full backtrace?

**Comment by tsahara** [comment]
You need `enable_debug` in your build_config.rb (though I don't know why :).  If you build mruby with `run_test.rb`, it should look like:

```rb
...

  conf.gem :git => 'https://github.com/iij/mruby-env.git'
  conf.enable_test
  enable_debug

  conf.gem File.expand_path(File.dirname(__FILE__))
end
```

And if you want fine-grained error reporting, import [iij/mruby-errno](https://github.com/iij/mruby-errno) mrbgem too.  It provides human readable error messages.  To import iij/mruby-errno, add a line:

```
    conf.gem :github => 'iij/mruby-errno'
```

**Comment by gfx** [comment]
Thank you for the information. I'll give it a try for the modules.

However, I believe concise error messages in default build settings are absolutely important, so I prefer this PR instead of just `rb_sys_raise(pathname)`.

**Comment by tsahara** [comment]
Okay, I changed error message to include "open".  Thank you for the feedback!

**Comment by gfx** [comment]
Thank you!

**Pull request closed without merging**